### PR TITLE
Migrate to a new PWD API

### PR DESCRIPTION
**Pull request by YizhePKU**
This is the first PR towards migrating to a new `$env.PWD` API that returns potentially un-canonicalized paths. Refer to PR #12515 for motivations.

## New API: `EngineState::cwd()`

The goal of the new API is to cover both parse-time and runtime use case, and avoid unintentional misuse. It takes an `Option<Stack>` as argument, which if supplied, will search  for `$env.PWD` on the stack in additional to the engine state. I think with this design, there's less confusion over parse-time and runtime environments. If you have access to a stack, just supply it; otherwise supply `None`.

## Deprecation of other PWD-related APIs

Other APIs are re-implemented using `EngineState::cwd()` and properly documented. They're marked deprecated, but their behavior is unchanged. Unused APIs are deleted, and code that accesses `$env.PWD` directly without using an API is rewritten.

Deprecated APIs:

* `EngineState::current_work_dir()`
* `StateWorkingSet::get_cwd()`
* `env::current_dir()`
* `env::current_dir_str()`
* `env::current_dir_const()`
* `env::current_dir_str_const()`

Other changes:

* `EngineState::get_cwd()` (deleted)
* `StateWorkingSet::list_env()` (deleted)
* `repl::do_run_cmd()` (rewritten with `env::current_dir_str()`)

## `cd` and `pwd` now use logical paths by default

This pulls the changes from PR #12515. It's currently somewhat broken because using non-canonicalized paths exposed a bug in our path normalization logic (Issue #12602). Once that is fixed, this should work.

## Future plans

This PR needs some tests. Which test helpers should I use, and where should I put those tests?

I noticed that unquoted paths are expanded within `eval_filepath()` and `eval_directory()` before they even reach the `cd` command. This means every paths is expanded twice. Is this intended?

Once this PR lands, the plan is to review all usages of the deprecated APIs and migrate them to `EngineState::cwd()`. In the meantime, these usages are annotated with `#[allow(deprecated)]` to avoid breaking CI.

### CODE REVIEW

File: crates/nu-protocol/src/engine/engine_state.rs
Comment (kubouch) [inline crates/nu-protocol/src/engine/engine_state.rs:967]: I think it would be better to return a String which users can convert to PathBuf if they need to. Otherwise, there will be unnecessary PathBuf->String conversion if string is required.

File: crates/nu-engine/src/env.rs
Comment (kubouch) [inline crates/nu-engine/src/env.rs:186]: Rather than canonicalizing with `"."`, `fs::canonicalize()` should be called with a comment that we canonicalize a guaranteed absolute path.

File: crates/nu-engine/src/env.rs
Comment (kubouch) [inline crates/nu-engine/src/env.rs:198]: Same as above

**Review by kubouch — CHANGES_REQUESTED** [review]

**Comment by kubouch** [comment]
Looks good! One issue is that `EngineState::cwd()` returns potentially logical path. IMO, it should return a fully resolved physical path, and we would have `EngineState::cwd_logical()` that can contain symlinks. Alternatively, we could have the current `cwd()` and add `cwd_physical()`, or be super explicit and have `cwd_logical()` + `cwd_physical()`. The point is to allow explicitly fetching the physical path, which we've relied on so far. We can then gradually move towards using the logical path later, but this approach minimizes the potential risks.

About tests: You can look around the codebase and see if there is something related already. `tests/path` might be ok, or since EngineState is in nu-protocol, you can add cwd tests there. The `Playground` in `nu-test-support` should be good for testing the current directory, you can set it to start in a provided cwd. It's used all over the place, you should find relevant examples in tests for filesystem commands.

Do you plan looking at https://github.com/nushell/nushell/issues/12602 as well?

Otherwise, I think we're getting close. I'll convert to draft to avoid merging before #12602 is fixed, but feel free to reopen once it's ready.

File: crates/nu-engine/src/env.rs
Comment (YizhePKU) [reply crates/nu-engine/src/env.rs:186]: We can't actually use `fs::canonicalize()` here because we need to call `to_winuser_path()` on Windows. I'd love to use `nu_path::canonicalize()` but that API is not exposed to public. So maybe using `canonicalize_with()` is OK? I'll add a comment explaining this situation.

File: crates/nu-protocol/src/engine/engine_state.rs
Comment (YizhePKU) [reply crates/nu-protocol/src/engine/engine_state.rs:967]: I'm not sure. Conceptually, "current working directory" should be a path, not a string. It also matches the behavior of `std::env::current_dir()`. Zero-cost conversion from `PathBuf` to `String` can be done with `path_buf.into_os_string().into_string().unwrap()`, but that involves an `unwrap()`.

On the other hand, returning a string does make some use cases more convenient. Maybe I'll add a `cwd_as_string()` method? Just one more API won't hurt...?

**Comment by YizhePKU** [comment]
> One issue is that EngineState::cwd() returns potentially logical path. IMO, it should return a fully resolved physical path, and we would have EngineState::cwd_logical() that can contain symlinks. [...] The point is to allow explicitly fetching the physical path, which we've relied on so far.

The idea is that users can resolve logical path into physical path, but not the other way around. Explicitly fetching the physical path can be done either by using the old APIs (`env::current_dir()` and friends) or by calling `canonicalize()` after fetching the logical path.

We could add an API that always returns physical path, but that only serves as a shortcut for "`EngineState::cwd()` then `canonicalize()`". I think we should hold back until we know how useful such a shortcut is.

> Do you plan looking at https://github.com/nushell/nushell/issues/12602 as well?

Yes. The fix should be simple -- just delete the early-exit part of the code. Testing path equality should be done by converting path to string.

**Comment by YizhePKU** [comment]
I've made PR #12662 to fix the path normalization bug. I'm ready to merge this once that PR lands.

File: crates/nu-protocol/src/engine/engine_state.rs
Comment (kubouch) [reply crates/nu-protocol/src/engine/engine_state.rs:965]: Yeah, one more API call is fine, that's why I made `env::current_dir()` and `env::current_dir_str()`. I'd try to avoid the `unwrap()`.

File: crates/nu-engine/src/env.rs
Comment (kubouch) [reply crates/nu-engine/src/env.rs:189]: `nu-path::canonicalize()` uses `Path.canonicalize()` which is an alias to `fs::canonicalize()`. What is the problem with `to_winuser_path()`? I'm not familiar with it.

In Nushell, calling `fs::canonicalize()` is OK _after the path has been expanded into an absolute path_`. That's why we always use `nu-path::canonicalize_with()` with the cwd coming from `$env.PWD` which should be guaranteed to be an absolute path, therefore the underlying `fs:canonicalize()` shouldn't use the wrong current dir from `std::env`.

File: crates/nu-protocol/src/engine/engine_state.rs
Comment (YizhePKU) [reply crates/nu-protocol/src/engine/engine_state.rs:965]: I think I'll add the API later while migrating the old API usages. It should give me a better idea whether an extra API is needed.

File: crates/nu-engine/src/env.rs
Comment (YizhePKU) [reply crates/nu-engine/src/env.rs:189]: On Windows, `fs::canonicalize()` will return UNC paths, which looks like `\\?\C:\Users`. This is considered "canonical", but not very user friendly. `nu-path::canonicalize()` is a wrapper over `fs::canonicalize()` that converts UNC paths into "legacy" paths like `C:\Users` on Windows (on other platforms it does nothing).

Calling `fs::canonicalize()` is incorrect here because we would get UNC paths. That's why I said "I'd love to use `nu-path::canonicalize()`", because that also performs the conversion without supplying a meaningless cwd.

Comment (YizhePKU) [reply crates/nu-engine/src/env.rs:189]: Side note: for the "UNC to legacy conversion" functionality, we're currently using the `omnipath` crate. However, the `dunce` crate is a lot more popular, has reached version `1.0`, and offers an API similar to our wrapper. It would be a good idea to use that instead.

Comment (ChrisDenton) [reply crates/nu-engine/src/env.rs:189]: The reason for the creation of `omnipath` was because the `dunce` crate has issues. E.g.: https://github.com/nushell/nushell/issues/6461#issuecomment-1234793374

If the dunce crate has since fixed them then great! Otherwise using it would be a regression.

Comment (kubouch) [reply crates/nu-engine/src/env.rs:189]: OK, I see it now in the code, I should have looked properly. Then, yes, let's use `nu-path::canonicalize()`.  The reason why it's not a public API is to not encourage people to use it and use `canonicalize_with()` instead. But we can make it public. Maybe it can be renamed to something slightly worrying, like `canonicalize_unsafe()`, with a docstring to discourage its usage unless you know what you're doing, just to minimize potential misuse.

@ChrisDenton The issue and the linked dunce issue are both closed. Seems fixed now?

Comment (ChrisDenton) [reply crates/nu-engine/src/env.rs:189]: The nushell issue was closed by switching to omnipath. The dunce issue looks like it was just closed without a fix. Maybe wontfix? I can't tell but it has been years at this point.

Comment (YizhePKU) [reply crates/nu-engine/src/env.rs:189]: I tested `dunce` and it doesn't seem to have been fixed.

On the other hand, `omnipath`'s approach is to just strip UNC prefix unconditionally, which is also incorrect. It might be the lesser evil for now. I'll keep looking.

Comment (YizhePKU) [reply crates/nu-engine/src/env.rs:189]: This test case fails:

```rust
#[test]
fn long_verbatim_path() {
    let long_filename = str::repeat("A", 500);
    let path = Path::new(r"\\?\C:\").join(long_filename);

    let result = path.to_winuser_path().unwrap();
    // `path` is over the 260 character limit, so we can't simplify it
    assert_eq!(result, path);
}
```

I looked at the code, and indeed `omnipath` tries to use `GetFullPathNameW` to check if stripping prefix is safe. This test case fails because it skips the check when the prefix is a drive letter. I think this is a bug?

Also, while I was testing, I noticed `GetFullPathNameW` was accepting long paths even though I never enabled it on my machine. Would you try this test case and tell me if it works on your machine?

```rust
#[test]
fn GetFullPathNameW_long_path() {
    let long_filename = PathBuf::from(str::repeat("A", 1000));
    let mut buffer = [0u16; 1200];
    let ret = unsafe {
        GetFullPathNameW(
            to_wide(&long_filename).unwrap().as_ptr(),
            1200,
            buffer.as_mut_ptr(),
            ptr::null_mut(),
        )
    };
    // If `GetFullPathNameW` accepts long paths, this should pass.
    assert!(ret >= 1000);
}
```

Comment (ChrisDenton) [reply crates/nu-engine/src/env.rs:189]: Ah, that's intentional. It's meant for displaying paths to the user. Using a path in filesystem APIs should automatically convert to the `\\?\` form if necessary.

I can add a function that does it only for short paths. Though I'd ask the use case? E.g. [`SetCurrentDirectory`](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-setcurrentdirectory#remarks) does not support `\\?\` paths. You can technically use such a path but then other APIs break because they assume `SetCurrentDirectory` is using either `C:\` or `\\server\share` style paths (as documented).

In any case it's risky to use paths longer than `MAX_PATH` in `SetCurrentDirectory` because it causes `CreateProcess` to fail, even if long paths are supported. I fairly recently added a warning box about that to the `SetCurrentDirectory` docs.

Comment (YizhePKU) [reply crates/nu-engine/src/env.rs:189]: I'm a bit confused. Are you saying the result of `to_winuser_path()` should not be used in Windows APIs? Because I think Nushell is currently doing exactly that.

Comment (YizhePKU) [reply crates/nu-engine/src/env.rs:189]: Just for clarification, when exactly does `to_winuser_path()` remove the verbatim prefix?

Comment (ChrisDenton) [reply crates/nu-engine/src/env.rs:188]: It always tries to. It only doesn't if doing so would change the path such that you couldn't recover the original verbatim path. For example, if the file name is `NUL` or if any component ends with a dot.

It does this by making sure `GetFullPathName` doesn't change the path.

**Comment by YizhePKU** [comment]
Merging from main breaks completion. Not sure what's the matter yet.

**Comment by YizhePKU** [comment]
So, about the `canonicalize_with()` problem: I actually tried exposing `nu_path::canonicalize()` as a public API, but then I felt silly for adding a public API just to make some deprecated code looks nicer. I think the comment I added should suffice for now? If we actually need `nu_path::canonicalize()` in some non-deprecated parts of the codebase, we can add it later.

**Review by kubouch — APPROVED** [review]

**Comment by kubouch** [comment]
Yeah, seems good. I somehow missed that it's in a deprecated code, it's fine.

**Comment by kubouch** [comment]
Seems like there is a merge conflict now. We can land it whenever it's fixed.

**Comment by YizhePKU** [comment]
I've fixed the merge conflict.

**Comment by kubouch** [comment]
OK, thanks!

**Comment by groundedtako** [comment]
I'm so psyched about this release!

**Pull request merged**
